### PR TITLE
RPC requests with trace id and span id

### PR DESCRIPTION
**Pull request by SimonRichardson**
The following implements the w3c tracing context document[1] for both trace-id[2] and span-id[3]. Sometimes span-id is called parent-id, but for out setup we're going to be explicit and call it span-id like the document allows.

Currently, there is no validation of the values of a trace-id or a span-id. At the moment, I'm thinking we can ignore that and force the OTEL (open telemetry) clients force that validation. Most likely, they'll be the ones seeding that information.

As tracing is optional, I don't want to bring down Juju for a bad trace and span id.

 1. https://www.w3.org/TR/2021/REC-trace-context-1-20211123/
 2. https://www.w3.org/TR/2021/REC-trace-context-1-20211123/#trace-id
 3. https://www.w3.org/TR/2021/REC-trace-context-1-20211123/#parent-id

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing

## QA steps

There should be no breakages, as this isn't fully wired up, there will be no difference.

```sh
$ juju bootstrap lxd test --build-agent
```

## Links

**Jira card:** JUJU-[XXXX]

### CODE REVIEW

File: api/apiclient.go [hunk 1]
```diff
@@ -1247,7 +1247,7 @@ func isX509Error(err error) bool {
 // object id, and the specific RPC method. It marshalls the Arguments, and will
 // unmarshall the result into the response object that is supplied.
 func (c *conn) APICall(facade string, vers int, id, method string, args, response interface{}) error {
-	err := c.client.Call(rpc.Request{
+	err := c.client.Call(context.TODO(), rpc.Request{
```
Comment (SimonRichardson) [inline api/apiclient.go:1250]: This will come from APICall, but to do that is about 1 million breaking changes, so let's do that in another PR.

File: rpc/server.go [hunk 2]
```diff
@@ -59,18 +59,26 @@ type Header struct {
 	// Request holds the action to invoke.
 	Request Request
 
-	// Error holds the error, if any.
+	// Error holds the error, if interface{}.
```
Comment (hpidcock) [inline rpc/server.go:62]: heh... looks like a bit of a find replace mistake here.

File: rpc/jsoncodec/codec.go [hunk 3]
```diff
@@ -74,21 +75,22 @@ type outMsgV1 struct {
...
 	ErrorInfo map[string]interface{} `json:"error-info,omitempty"`
 	Response  interface{}            `json:"response,omitempty"`
+	TraceID   string                 `json:"trace-id,omitempty"`
+	SpanID    string                 `json:"span-id,omitempty"`
 }
 
+// Close closes the underlying connection and sets the codec to
+// closing mode, so that interface{} further errors are ignored.
 func (c *Codec) Close() error {
-	c.mu.Lock()
```
Comment (hpidcock) [inline rpc/jsoncodec/codec.go:80]: Probably just need to keep an eye on the race tests as there might of been something hanging off these fences.

File: rpc/server.go [hunk 4]
```diff
@@ -229,7 +237,7 @@ func (conn *Conn) Start(ctx context.Context) {
 // with specified codes.  There will be a panic if transformErrors
 // returns nil.
 //
-// Serve may be called at any time on a connection to change the
+// Serve may be called at interface{} time on a connection to change the
```
Comment (hpidcock) [inline rpc/server.go:240]: "interface{}time, interface{} place, interface{}where!"

**Review by hpidcock — APPROVED** [review]
